### PR TITLE
Fix TiKV Config Incompatibility error

### PR DESCRIPTION
**Pull request by Yisaer**
Close https://github.com/pingcap/tidb-operator/issues/2592

Add a function that transform the generated tikv ConfigMap templates which dependes on the tikv Version.

Here is the example for tikv 4.0 and 3.1:

```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic2
spec:
  #  version: latest
  timezone: UTC
  enableTLSCluster: true
  pvReclaimPolicy: Delete
  imagePullPolicy: IfNotPresent
  pd:
    image: pingcap/pd:v4.0.0
    replicas: 1
    requests:
      storage: "1Gi"
    config: {}
    storageClassName: local-storage
    service:
      type: NodePort
  tikv:
    config:
      pessimistic-txn:
        wait-for-lock-timeout: "1s"
        wake-up-delay-duration: "20ms"
    image: pingcap/tikv:v4.0.0
    replicas: 1
    requests:
      storage: "1Gi"
    storageClassName: local-storage
  tidb:
    enableAdvertiseAddress: true
    image: pingcap/tidb:v4.0.0
    imagePullPolicy: IfNotPresent
    replicas: 1
    service:
      type: NodePort
    config: {}
    storageClassName: local-storage
    requests:
      cpu: 1
```

```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic
spec:
  #  version: latest
  timezone: UTC
  enableTLSCluster: true
  pvReclaimPolicy: Delete
  imagePullPolicy: IfNotPresent
  pd:
    image: pingcap/pd:v3.1.1
    replicas: 1
    requests:
      storage: "1Gi"
    config: {}
    storageClassName: local-storage
    service:
      type: NodePort
  tikv:
    config:
      pessimistic-txn:
        wait-for-lock-timeout: "1000"
        wake-up-delay-duration: "20"
    image: pingcap/tikv:v3.1.1
    replicas: 1
    requests:
      storage: "1Gi"
    storageClassName: local-storage
  tidb:
    enableAdvertiseAddress: true
    image: pingcap/tidb:v3.1.1
    imagePullPolicy: IfNotPresent
    replicas: 1
    service:
      type: NodePort
    config: {}
    storageClassName: local-storage
    requests:
      cpu: 1
```

### Does this PR introduce a user-facing change?:

```release-note
Fix the `TikvConfig` for `wait-for-lock-timeout` and `wake-up-delay-duration`  won't take effect to the `TiKV` version under 4.0.0 
```

### CODE REVIEW

File: pkg/manager/member/tikv_member_manager.go [hunk 1]
```diff
@@ -511,6 +511,37 @@ func volumeClaimTemplate(r corev1.ResourceRequirements, metaName string, storage
 	}
 }
 
+func transformTiKVConfigMap(src []byte, tc *v1alpha1.TidbCluster) string {
```
Comment (cofyc) [inline pkg/manager/member/tikv_member_manager.go:514]: change `[]byte` to the string, it's better we keep the same type for the input & output and only do the transformation

File: pkg/manager/member/tikv_member_manager.go [hunk 2]
```diff
@@ -511,6 +511,37 @@ func volumeClaimTemplate(r corev1.ResourceRequirements, metaName string, storage
...
+	config := tc.Spec.TiKV.Config
+	if config == nil {
+		return string(src)
+	}
+	srcStr := string(src)
+	if config.TiKVPessimisticTxn != nil {
+		clusterVersionGE4, err := clusterVersionGreaterThanOrEqualTo4(tc.TiKVVersion())
+		if err != nil {
+			klog.Infof("cluster version: %s is not semantic versioning compatible", tc.TiKVVersion())
```
Comment (cofyc) [inline pkg/manager/member/tikv_member_manager.go:523]: we assume the version is greater or equal to 4 on error, so return the configuration here and we don't need to check error later

File: pkg/manager/member/tikv_member_manager_test.go [hunk 3]
```diff
@@ -2074,3 +2074,77 @@ func TestGetTiKVConfigMap(t *testing.T) {
...
+			result: `[pessimistic-txn]
+  wait-for-lock-timeout = "1s"
+  wake-up-delay-duration = "20ms"
+`,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tc.Spec.TiKV.Image = fmt.Sprintf("tikv:%d.0.0", test.version)
```
Comment (cofyc) [inline pkg/manager/member/tikv_member_manager_test.go:2107]: create a new tidb cluster in each test

File: pkg/manager/member/tikv_member_manager_test.go [hunk 4]
```diff
@@ -2074,3 +2074,77 @@ func TestGetTiKVConfigMap(t *testing.T) {
 		})
 	}
 }
+
+func TestTransformTiKVConfigMap(t *testing.T) {
+	g := NewGomegaWithT(t)
+	tc := newTidbClusterForTiKV()
+	type testcase struct {
+		name    string
+		version int
```
Comment (cofyc) [inline pkg/manager/member/tikv_member_manager_test.go:2083]: add a field for `v1alpha1.TiKVPessimisticTxn`, the content of this configuration is part of test case

File: pkg/manager/member/tikv_member_manager.go (hunk 1 above)
Comment (cofyc) [inline pkg/manager/member/tikv_member_manager.go:514]: can we also add a link to tikv change PR on these fields? then people can know why this function is required

Comment (Yisaer) [reply pkg/manager/member/tikv_member_manager.go]: updated.

File: pkg/manager/member/tikv_member_manager_test.go (hunk 4 above)
Comment (Yisaer) [reply pkg/manager/member/tikv_member_manager_test.go]: updated

File: pkg/manager/member/tikv_member_manager.go (hunk 1 above)
Comment (Yisaer) [reply pkg/manager/member/tikv_member_manager.go]: updated.

**Comment by cofyc** [comment]
btw, can you create an issue for `MarshalTOML` solution?

**Comment by Yisaer** [comment]
> btw, can you create an issue for `MarshalTOML` solution?

sure

File: pkg/manager/member/tikv_member_manager.go [hunk 5]
```diff
@@ -511,6 +512,38 @@ func volumeClaimTemplate(r corev1.ResourceRequirements, metaName string, storage
...
+	if config.TiKVPessimisticTxn != nil {
+		if config.TiKVPessimisticTxn.WaitForLockTimeout != nil {
+			_, err := strconv.ParseInt(*config.TiKVPessimisticTxn.WaitForLockTimeout, 10, 64)
+			if err == nil {
+				waitForLockTimeOutKey := "wait-for-lock-timeout"
+				old := fmt.Sprintf(`%s = "%s"`, waitForLockTimeOutKey, *config.TiKVPessimisticTxn.WaitForLockTimeout)
+				newString := fmt.Sprintf(`%s = %s`, waitForLockTimeOutKey, *config.TiKVPessimisticTxn.WaitForLockTimeout)
+				srcStr = strings.ReplaceAll(srcStr, old, newString)
+			}
+			if config.TiKVPessimisticTxn.WakeUpDelayDuration != nil {
```
Comment (zjj2wry) [inline pkg/manager/member/tikv_member_manager.go:533]: we only need to convert the data type of config.TiKVPessimisticTxn.WakeUpDelayDuration when `if config.TiKVPessimisticTxn.WaitForLockTimeout != nil`? what happen when `config.TiKVPessimisticTxn.WaitForLockTimeout != nil` and `config.TiKVPessimisticTxn.WakeUpDelayDuration != nil` ？

**Comment by zjj2wry** [comment]
@Yisaer I understand that this pr will convert string to int, what will happen if users use tidb-3.1  image, current will test different version images？

File: pkg/manager/member/tikv_member_manager.go (hunk 5 above)
Comment (Yisaer) [reply pkg/manager/member/tikv_member_manager.go]: good catch. updated.

**Comment by Yisaer** [comment]
> @Yisaer I understand that this pr will convert string to int, what will happen if users use tidb-3.1 image, current will test different version images？

The transforming depends on the configuration content type. If it's numeric ,like `wait-for-lock-timeout: "1000"`, it would be rendered as `wait-for-lock-timeout: 1000` and can be accptable by tikv 3.1 and 4.0

**Pull request merged**